src/ConcreteOperations/cartesian_product.jl

```diff
@@ -180,3 +180,11 @@ function cartesian_product(H1::AbstractHyperrectangle, H2::AbstractHyperrectangl
     r = vcat(radius_hyperrectangle(H1), radius_hyperrectangle(H2))
     return Hyperrectangle(c, r)
 end
+
+@commutative function cartesian_product(H::AbstractHyperrectangle{N},
+                                        U::Universe) where {N}
+    n = dim(H) + dim(U)
+    constraints = [HalfSpace(SingleEntryVector(c.a.i, n, c.a.v), c.b) for c in
```

Comment:
this assumes that the `constraints_list` of an AbstractHyperrectangle *always* returns a SEV, which could not be the case in general. i see two alternatives:

- dispatch on the concrete type Hyperrectangle{..., SEV}, or
- (my preferred choice) use [iterated dispatch](https://docs.julialang.org/en/v1/manual/methods/#Iterated-dispatch) along these lines:

```julia
@commutative function cartesian_product(H::AbstractHyperrectangle{N},
                                        U::Universe) where {N}
    n = dim(H) + dim(U)
    clist = constraints_list(H)
    return _cartesian_product_hyperrectangle_universe(clist, U)
end

function _cartesian_product_hyperrectangle_universe(clist::Vector{SingleEntryVector{N}}, U) where {N}
    constraints = [HalfSpace(SingleEntryVector(c.a.i, n, c.a.v), c.b) for c in clist]
    return HPolyhedron(constraints)
end
```

Reply:
I improved it quite a lot. Going down to the `HalfSpace` generalizes to arbitrary polyhedra. Before we went through `Polyhedra` in that case. Now we do not and if the polyhedron has sparse constraints, so will the result of `cartesian_product`.

The idea is to either append or prepend zeros to the normal direction, depending on whether the universe comes last or first.